src/command/lsp/mod.rs

```diff
@@ -0,0 +1,230 @@
+            .setup_composition_watcher(
+                lazily_resolved_supergraph_config
+                    .fully_resolve_subgraphs(&client_config, &studio_client)
+                    .await
+                    .map_err(ResolveSupergraphConfigError::ResolveSubgraphs)?,
+                supergraph_binary,
+                TokioCommand::default(),
+                FsReadFile::default(),
+                FsWriteFile::default(),
+                OutputTarget::Stdout,
+                Utf8PathBuf::try_from(temp_dir())?,
+            )
+            .run();
+
+        let supergraph_yaml_url =
+            tower_lsp::lsp_types::Url::from_file_path(supergraph_yaml_path)
+                .map_err(|_| anyhow!("Failed to convert supergraph yaml path to url"))?;
+
+        while let Some(event) = stream.next().await {
+            match event {
+                CompositionEvent::Started => {
+                    language_server.composition_did_start().await;
+                }
+                CompositionEvent::Success(CompositionSuccess {
+                    supergraph_sdl,
+                    federation_version,
+                    hints,
+                }) => {
+                    debug!("Successfully composed with version {}", federation_version);
+                    // Clear any previous errors on `supergraph.yaml`
+                    language_server
+                        .publish_diagnostics(supergraph_yaml_url.clone(), vec![])
+                        .await;
+                    language_server
+                        .composition_did_update(
+                            Some(supergraph_sdl),
+                            hints.into_iter().map(Into::into).collect(),
+                            None,
+                        )
+                        .await;
+                }
+                CompositionEvent::Error(CompositionError::Build { source: errors }) => {
+                    debug!(
+                        ?errors,
+                        "Composition {federation_version} completed with errors"
+                    );
+                    // Clear any previous errors on `supergraph.yaml`
+                    language_server
+                        .publish_diagnostics(supergraph_yaml_url.clone(), vec![])
+                        .await;
+                    language_server
+                        .composition_did_update(
+                            None,
+                            errors.into_iter().map(Into::into).collect(),
+                            None,
+                        )
+                        .await
+                }
+                CompositionEvent::Error(err) => {
+                    debug!("Composition {federation_version} failed: {err}");
+                    // TODO: we could highlight the version of federation, since it failed.
+                    let message = format!("Failed run composition {federation_version}: {err}",);
```

Comment:
Not sure what this comment means, it came over verbatim from the original PR

Reply:
Basically instead of using `Range::default()` below, which results in no red squigglies in any files, if we start retaining location data when we parse `supergraph.yaml`, we could highlight part of that file when this error happens. The version is the most likely place, since it means there was something wrong with running the composition binary, not something wrong with actually composing.

Reply:
Ah! I get what you mean, do we just want to leave this as a TODO then? Or do you want to make this change as we integrate?

Reply:
I don't think it's critical, I had it as a TODO in case I got time to improve it but.. clearly I didn't 😛. I think it'd require changing `apollo-federation-types` to use a parsing library which retains location data (instead of `serde_yaml` or whatever), so probably more of a change than we want unless a customer pushes for it.

In short, just delete the TODO is probably fine